lib/trino-hive-formats/src/test/java/io/trino/hive/formats/FormatTestUtils.java

```diff
@@ -0,0 +1,600 @@
+    }
+
+    public static void assertColumnValueEquals(Type type, Object actual, Object expected)
+    {
+        if (actual == null) {
+            assertNull(expected);
+            return;
+        }
+        if (type instanceof ArrayType) {
+            List<?> actualArray = (List<?>) actual;
+            List<?> expectedArray = (List<?>) expected;
+            assertEquals(actualArray.size(), expectedArray.size());
+
+            Type elementType = type.getTypeParameters().get(0);
+            for (int i = 0; i < actualArray.size(); i++) {
+                Object actualElement = actualArray.get(i);
+                Object expectedElement = expectedArray.get(i);
+                assertColumnValueEquals(elementType, actualElement, expectedElement);
+            }
+        }
+        else if (type instanceof MapType) {
+            Map<?, ?> actualMap = (Map<?, ?>) actual;
+            Map<?, ?> expectedMap = (Map<?, ?>) expected;
+            assertEquals(actualMap.size(), expectedMap.size());
+
+            Type keyType = type.getTypeParameters().get(0);
+            Type valueType = type.getTypeParameters().get(1);
+
+            List<Entry<?, ?>> expectedEntries = new ArrayList<>(expectedMap.entrySet());
+            for (Entry<?, ?> actualEntry : actualMap.entrySet()) {
+                Iterator<Entry<?, ?>> iterator = expectedEntries.iterator();
+                while (iterator.hasNext()) {
+                    Entry<?, ?> expectedEntry = iterator.next();
+                    try {
+                        assertColumnValueEquals(keyType, actualEntry.getKey(), expectedEntry.getKey());
+                        assertColumnValueEquals(valueType, actualEntry.getValue(), expectedEntry.getValue());
+                        iterator.remove();
+                    }
+                    catch (AssertionError ignored) {
+                    }
+                }
+            }
+            assertTrue(expectedEntries.isEmpty(), "Unmatched entries " + expectedEntries);
+        }
+        else if (type instanceof RowType) {
+            List<Type> fieldTypes = type.getTypeParameters();
+
+            List<?> actualRow = (List<?>) actual;
+            List<?> expectedRow = (List<?>) expected;
+            assertEquals(actualRow.size(), fieldTypes.size());
+            assertEquals(actualRow.size(), expectedRow.size());
+
+            for (int fieldId = 0; fieldId < actualRow.size(); fieldId++) {
+                Type fieldType = fieldTypes.get(fieldId);
+                Object actualElement = actualRow.get(fieldId);
+                Object expectedElement = expectedRow.get(fieldId);
+                assertColumnValueEquals(fieldType, actualElement, expectedElement);
+            }
+        }
+        else if (type.equals(DOUBLE)) {
+            Double actualDouble = (Double) actual;
+            Double expectedDouble = (Double) expected;
+            if (actualDouble.isNaN()) {
```

Comment:
This doesn't seem needed. The TestNG `assertEquals(double, double, double)` method handles NaN explicitly.